docs: add release notes and new doc page about new application config (#1634)
commercetools/merchant-center-application-kit: website/src/releases/2020-07/application-config.mdx

---
date: 2020-07-16
title: Introducing a new and simpler application config
description: |
  We introduced a new configuration file that aims to simplify the way Custom Application are configured.\
  This also marks the deprecation of the `env.json` and `headers.json` files.
type: feature
topics:
  - Configuration
---

This release introduces the usage of a new configuration file format and marks the deprecation of the `env.json` and `headers.json` files.

<Info>

The `env.json` and `headers.json` files will still keep working but they will be removed in the next major release.

</Info>

> The new configuration format aims to drastically simplify how to configure the Custom Application, as well as to make the configuration process less error prone.

Reviewer: ```suggestion
The new configuration format aims to drastically simplify the configuration of a Custom Application. In addition, it also strives to make the configuration process less error prone.
```

> In fact, the new configuration file is backed by a JSON schema that is shipped together with the [new package](https://www.npmjs.com/package/@commercetools-frontend/application-config). The configuration file is then validated against the JSON schema.

Reviewer: ```suggestion
To achieve this, the new configuration file is backed by a JSON schema that is shipped together with the [new package](https://www.npmjs.com/package/@commercetools-frontend/application-config). The configuration file is then validated against the JSON schema.
```

> Furthermore, the new configuration file tries to infer as many required information as possible.

Reviewer: ```suggestion
Furthermore, the new configuration process tries to infer as much information as possible to reduce the amount of required fields.
```

<!--more-->

# Migrating to the new configuration file format

The new configuration file is a JSON file with one of the following names:

- `.custom-application-configrc`
- `.custom-application-config.json`
- `custom-application-config.json`

> The file is automatically loaded by the other packages, so you don't need to explicitly specify it in, for example, the `mc-scripts` commands.

Reviewer: ```suggestion
The file is automatically loaded by the packages depending on it, so you don't need to explicitly specify it anywhere. This applies for instance to the `mc-scripts` commands.
```

<Info>

> The file needs to be located in the project root folder.

Reviewer: ```suggestion
The file needs to be present in the project root folder.
```

  }
}
```

and for production mode `env.prod.json` and `headers.prod.json`:

```json title="env.prod.json"
{
  "applicationName": "Avengers app",
  "frontendHost": "avengers.app",
  "mcApiUrl": "https://mc-api.europe-west1.gcp.commercetools.com",
  "location": "gcp-eu",
  "env": "production",
  "cdnUrl": "https://cdn.avengers.app",
  "servedByProxy": true
}
```

```json title="headers.prod.json"
{
  "csp": {
    "script-src": ["avengers.app", "cdn.avengers.app"],
    "connect-src": [
      "mc-api.europe-west1.gcp.commercetools.com",
      "avengers.app",
      "cdn.avengers.app"
    ],
    "style-src": ["avengers.app", "cdn.avengers.app"]
  }
}
```

To migrate them to the new format, add a `custom-application-config.json` (or one of the other file names) with the following content:

```json title="custom-application-config.json"
{
  "name": "Avengers app",
  "entryPointUriPath": "avengers",
  "cloudIdentifier": "gcp-eu",
  "env": {
    "production": {
      "url": "https://avengers.app",
      "cdnUrl": "https://cdn.avengers.app"
    }
  }
}
```

That's it! All other values are inferred from the config, like the Content-Security-Policy (CSP) headers, etc.

<Info>

> Note that the environment placeholder values `${env:VALUE}` are still working.

Reviewer: ```suggestion
Additionally, note that the environment placeholder syntax `${env:VALUE}` continues to work.
```

</Info>

# Migrating `mc-scripts` commands

> The new configuration file is automatically loaded and therefore there is no need to explicitly pass the file path to the `mc-scripts` commands in the `package.json`

Reviewer: ```suggestion
The new configuration file is automatically loaded. Therefore, there is no need to explicitly pass the file path to the `mc-scripts` commands in the `package.json`
```

  }
}
```

```diff title="Updating the docker command to start the production server"
docker run \
  -v $(pwd):/etc/app \
  -p 8080:8080 \
-  eu.gcr.io/ct-images/mc-http-server \
+  eu.gcr.io/ct-images/mc-http-server
-  --config /etc/app/env.json \
-  --headers /etc/app/headers.json
```

```diff title="Updating the compile-html command"
{
  "scripts": {
-    "compile-html": "mc-scripts compile-html --headers=$(pwd)/headers.prod.json --config=$(pwd)/env.prod.json --use-local-assets --transformer $(pwd)/transformer-vercel.js"
+    "compile-html": "mc-scripts compile-html --use-local-assets --transformer $(pwd)/transformer-vercel.js"
  }
}
```

# JSON Schema support for VSCode

In the VSCode settings (either user settings or workspace settings), reference the schema JSON as described below:

```json title=".vscode/settings.json"
"json.schemas": [
  {
    "fileMatch": [
      "/.custom-application-configrc",
      "/.custom-application-config.json",
      "/custom-application-config.json"
    ],
    "url": "https://unpkg.com/@commercetools-frontend/application-config/schema.json"
  }
]
```

> Then your editor can offer autocompletion and validation of the JSON properties.

Reviewer: ```suggestion
Through this your editor can offer autocompletion and validation of the JSON properties.
```
